src/circuit_writer/mod.rs

```diff
@@ -226,7 +226,18 @@ impl<B: Backend> CircuitWriter<B> {
         }
 
         // compile function
-        circuit_writer.compile_main_function(fn_env, &function)?;
+        let returned_cells = circuit_writer.compile_main_function(fn_env, &function)?;
+        let main_span = circuit_writer.main_info().unwrap().span;
+        let private_input_indices = circuit_writer.private_input_indices.clone();
```

Comment:
I think `private_input_indices` should keep track of `CellVar`s instead of indicies, as indices is a backend implementation detail.

Maybe we should remove the `pub` in `CellVar { pub index: usize` and see what breaks